api/core/number.go

```diff
@@ -0,0 +1,539 @@
+	return rawPtrToUint64Ptr(n.AsRawPtr())
+}
+
+// - coerce
+
+func (n Number) CoerceToInt64(kind NumberKind) int64 {
+	switch kind {
+	case Int64NumberKind:
+		return n.AsInt64()
+	case Float64NumberKind:
+		return int64(n.AsFloat64())
+	case Uint64NumberKind:
+		return int64(n.AsUint64())
+	default:
+		// you get what you deserve
+		return 0
+	}
+}
+
+func (n Number) CoerceToFloat64(kind NumberKind) float64 {
+	switch kind {
+	case Int64NumberKind:
+		return float64(n.AsInt64())
+	case Float64NumberKind:
+		return n.AsFloat64()
+	case Uint64NumberKind:
+		return float64(n.AsUint64())
+	default:
+		// you get what you deserve
+		return 0
+	}
+}
+
+func (n Number) CoerceToUint64(kind NumberKind) uint64 {
+	switch kind {
+	case Int64NumberKind:
+		return uint64(n.AsInt64())
+	case Float64NumberKind:
+		return uint64(n.AsFloat64())
+	case Uint64NumberKind:
+		return n.AsUint64()
+	default:
+		// you get what you deserve
+		return 0
+	}
+}
+
+// - set
+
+func (n *Number) SetNumber(nn Number) {
+	*n.AsRawPtr() = nn.AsRaw()
+}
+
+func (n *Number) SetRaw(r uint64) {
+	*n.AsRawPtr() = r
+}
+
+func (n *Number) SetInt64(i int64) {
+	*n.AsInt64Ptr() = i
+}
+
+func (n *Number) SetFloat64(f float64) {
+	*n.AsFloat64Ptr() = f
+}
+
+func (n *Number) SetUint64(u uint64) {
+	*n.AsUint64Ptr() = u
+}
+
+// - set atomic
+
+func (n *Number) SetNumberAtomic(nn Number) {
+	atomic.StoreUint64(n.AsRawPtr(), nn.AsRaw())
+}
+
+func (n *Number) SetRawAtomic(r uint64) {
+	atomic.StoreUint64(n.AsRawPtr(), r)
+}
+
+func (n *Number) SetInt64Atomic(i int64) {
+	atomic.StoreInt64(n.AsInt64Ptr(), i)
+}
+
+func (n *Number) SetFloat64Atomic(f float64) {
+	atomic.StoreUint64(n.AsRawPtr(), float64ToRaw(f))
+}
+
+func (n *Number) SetUint64Atomic(u uint64) {
+	atomic.StoreUint64(n.AsUint64Ptr(), u)
+}
+
+// - swap
+
+func (n *Number) SwapNumber(nn Number) Number {
+	old := *n
+	n.SetNumber(nn)
+	return old
+}
+
+func (n *Number) SwapRaw(r uint64) uint64 {
+	old := n.AsRaw()
+	n.SetRaw(r)
+	return old
+}
+
+func (n *Number) SwapInt64(i int64) int64 {
+	old := n.AsInt64()
+	n.SetInt64(i)
+	return old
+}
+
+func (n *Number) SwapFloat64(f float64) float64 {
+	old := n.AsFloat64()
+	n.SetFloat64(f)
+	return old
+}
+
+func (n *Number) SwapUint64(u uint64) uint64 {
+	old := n.AsUint64()
+	n.SetUint64(u)
+	return old
+}
+
+// - swap atomic
+
+func (n *Number) SwapNumberAtomic(nn Number) Number {
+	return NewNumberFromRaw(atomic.SwapUint64(n.AsRawPtr(), nn.AsRaw()))
+}
+
+func (n *Number) SwapRawAtomic(r uint64) uint64 {
+	return atomic.SwapUint64(n.AsRawPtr(), r)
+}
+
+func (n *Number) SwapInt64Atomic(i int64) int64 {
+	return atomic.SwapInt64(n.AsInt64Ptr(), i)
+}
+
+func (n *Number) SwapFloat64Atomic(f float64) float64 {
+	return rawToFloat64(atomic.SwapUint64(n.AsRawPtr(), float64ToRaw(f)))
+}
+
+func (n *Number) SwapUint64Atomic(u uint64) uint64 {
+	return atomic.SwapUint64(n.AsUint64Ptr(), u)
+}
+
+// - add
+
+func (n *Number) AddNumber(kind NumberKind, nn Number) {
```

Comment:
This API should likely return a value being a sum, so we can atomically get an updated value in atomic versions.

Reply:
I see. I'd like to leave this for a future PR.